core/src/main/java/io/substrait/relation/ProtoRelConverter.java

```diff
@@ -0,0 +1,249 @@
+      }
+    }
+  }
+
+  private Rel newRead(ReadRel rel) {
+    if (rel.hasVirtualTable()) {
+      return newVirtualTable(rel);
+    } else if (rel.hasNamedTable()) {
+      return newNamedScan(rel);
+    } else {
+      return newEmptyScan(rel);
+    }
+  }
+
+  private Filter newFilter(FilterRel rel) {
+    var input = from(rel.getInput());
+    // TODO: For case of subquery, we'll need to pass in ProtoRelConverter to
+    //       ProtoExpressionConverter with nested structure of Rel/Expression.
+    return Filter.builder()
+        .input(input)
+        .condition(
+            new ProtoExpressionConverter(lookup, extensions, input.getRecordType())
+                .from(rel.getCondition()))
+        .remap(optionalRelmap(rel.getCommon()))
+        .build();
+  }
+
+  private NamedStruct newNamedStruct(ReadRel rel) {
+    var namedStruct = rel.getBaseSchema();
+    var struct = namedStruct.getStruct();
+    return ImmutableNamedStruct.builder()
+        .names(namedStruct.getNamesList())
+        .struct(
+            Type.Struct.builder()
+                .fields(struct.getTypesList().stream().map(FromProto::from).toList())
+                .nullable(FromProto.isNullable(struct.getNullability()))
+                .build())
+        .build();
+  }
+
+  private EmptyScan newEmptyScan(ReadRel rel) {
+    var namedStruct = newNamedStruct(rel);
+    return EmptyScan.builder()
+        .initialSchema(namedStruct)
+        .remap(optionalRelmap(rel.getCommon()))
+        .filter(
+            Optional.ofNullable(
+                rel.hasFilter()
+                    ? new ProtoExpressionConverter(lookup, extensions, namedStruct.struct())
+                        .from(rel.getFilter())
+                    : null))
+        .build();
+  }
+
+  private NamedScan newNamedScan(ReadRel rel) {
+    var namedStruct = newNamedStruct(rel);
+    return NamedScan.builder()
+        .initialSchema(namedStruct)
+        .names(rel.getNamedTable().getNamesList())
+        .remap(optionalRelmap(rel.getCommon()))
+        .filter(
+            Optional.ofNullable(
+                rel.hasFilter()
+                    ? new ProtoExpressionConverter(lookup, extensions, namedStruct.struct())
+                        .from(rel.getFilter())
+                    : null))
+        .build();
+  }
+
+  private VirtualTableScan newVirtualTable(ReadRel rel) {
+    var virtualTable = rel.getVirtualTable();
+    List<Expression.StructLiteral> structLiterals = new ArrayList<>(virtualTable.getValuesCount());
+    for (var struct : virtualTable.getValuesList()) {
+      structLiterals.add(
+          ImmutableExpression.StructLiteral.builder()
+              .fields(struct.getFieldsList().stream().map(ProtoExpressionConverter::from).toList())
+              .build());
+    }
+    var converter =
+        new ProtoExpressionConverter(lookup, extensions, newNamedScan(rel).getRecordType());
```

Comment:
This will create a namedStruct from the baseSchema. Is that the right type to pass through?

Reply:
I wouldn't expect this record type to get used. I'd probably pass in an empty struct type. If we try to index into it, I'd hope we'd throw an exception. There is no input to a virtual table so proto expression converter shouldn't need a root type in this case.

Reply:
I created a static EMPTY_TYPE and use that instead. I also enforce that rootType is non null.